default value as parameter failing

I have schema with default value in parameter as following example;

**Here is the input**
```
type Query {
  test(X: Int = 20): Int
}
```

when I call the library;
```
superSchema := gql.Merge(" ", "./test.graphql")
```

**Expected Result:**
Successfully parsing

**Actual**

```
panic: lib.Error("/Users/test/Development/repo/test.graphql:2:19: unexpected "20"")

goroutine 1 [running]:
github.com/mununki/gqlmerge/lib.errorf(...)
	/Users/test/go/pkg/mod/github.com/mununki/gqlmerge@v0.2.14/lib/parse.go:14
github.com/mununki/gqlmerge/lib.(*lexer).consumeIdent(0xc000130000, {0x0, 0x0, 0xc000134090?})
	/Users/test/go/pkg/mod/github.com/mununki/gqlmerge@v0.2.14/lib/lex.go:486 +0x2df
github.com/mununki/gqlmerge/lib.(*Parser).parseArgs(0xc00013dcb0)
	/Users/test/go/pkg/mod/github.com/mununki/gqlmerge@v0.2.14/lib/parse.go:73 +0x44a
```

--- a/lib/graphql.go
+++ b/lib/graphql.go
@@ -82,6 +82,7 @@ type Field struct {
 	Null         bool
 	IsList       bool
 	IsListNull   bool
+	DefaultValue *string
 	Directives   []*Directive
 	Descriptions *[]string
 	Comments     *[]string

--- a/lib/graphql.go
+++ b/lib/graphql.go
@@ -64,14 +64,14 @@ type Type struct {
 }
 
 type Arg struct {
-	Name         string
-	Type         string
-	DefaultValue *string // in case of default values e.g. admin(role: Role = ADMIN): Admin!
-	Null         bool
-	IsList       bool
-	IsListNull   bool
-	Directives   []*Directive
-	Descriptions *[]string
+	Name          string
+	Type          string
+	DefaultValues *[]string // in case of default values e.g. admin(role: Role = ADMIN): Admin!
+	Null          bool
+	IsList        bool
+	IsListNull    bool
+	Directives    []*Directive
+	Descriptions  *[]string
 }
 
 type Field struct {

--- a/lib/graphql.go
+++ b/lib/graphql.go
@@ -66,7 +66,7 @@ type Type struct {
 type Arg struct {
 	Name         string
 	Type         string
-	TypeExt      *string // in case of default values e.g. admin(role: Role = ADMIN): Admin!
+	DefaultValue *string // in case of default values e.g. admin(role: Role = ADMIN): Admin!
 	Null         bool
 	IsList       bool
 	IsListNull   bool

--- a/lib/graphql.go
+++ b/lib/graphql.go
@@ -66,7 +66,7 @@ type Type struct {
 type Arg struct {
 	Name         string
 	Type         string
-	TypeExt      *string // in case of enum e.g. admin(role: Role = ADMIN): Admin!
+	TypeExt      *string // in case of default values e.g. admin(role: Role = ADMIN): Admin!
 	Null         bool
 	IsList       bool
 	IsListNull   bool

--- a/lib/graphql.go
+++ b/lib/graphql.go
@@ -76,16 +76,16 @@ type Arg struct {
 
 type Field struct {
 	BaseFileInfo
-	Name         string
-	Args         []*Arg
-	Type         string
-	Null         bool
-	IsList       bool
-	IsListNull   bool
-	DefaultValue *string
-	Directives   []*Directive
-	Descriptions *[]string
-	Comments     *[]string
+	Name          string
+	Args          []*Arg
+	Type          string
+	Null          bool
+	IsList        bool
+	IsListNull    bool
+	DefaultValues *[]string
+	Directives    []*Directive
+	Descriptions  *[]string
+	Comments      *[]string
 }
 
 type Scalar struct {

--- a/lib/lex.go
+++ b/lib/lex.go
@@ -489,3 +489,38 @@ func (l *lexer) consumeIdent(includings ...tokenType) (*token, *[]string) {
 		return tok, &comments
 	}
 }
+
+/*
+The `consumeIdent` function treats tokString as a comment.
+If you want to scan the tokString for ident, you can use the consumeIdentInclString function.
+
+e.g. scan "woonki" for ident here
+
+test(name: String = "woonki"): Bool
+*/
+func (l *lexer) consumeIdentInclString(includings ...tokenType) (*token, *[]string) {
+	comments := []string{}
+	includings = append(includings, tokString)
+
+	for {
+		tok := l.next()
+		if tok.typ == tokSingleLineComment || tok.typ == tokBlockString {
+			comments = append(comments, tok.String())
+			continue
+		}
+
+		isIncluded := false
+		for _, incl := range includings {
+			if tok.typ == incl {
+				isIncluded = true
+				break
+			}
+		}
+
+		if tok.typ != tokIdent && !isIncluded {
+			errorf(`%s:%d:%d: unexpected "%s"`, l.filename, l.line, l.col, tok.String())
+		}
+		l.skipSpace()
+		return tok, &comments
+	}
+}

--- a/lib/parse.go
+++ b/lib/parse.go
@@ -79,7 +79,7 @@ func (p *Parser) parseArgs() []*Arg {
 					p.lex.consumeToken(tokEqual)
 					tex, _ := p.lex.consumeIdentInclString(tokNumber)
 					te := tex.String()
-					arg.TypeExt = &te
+					arg.DefaultValue = &te
 				}
 			}
 			arg.Directives = p.parseDirectives()

--- a/lib/parse.go
+++ b/lib/parse.go
@@ -68,7 +68,7 @@ func (p *Parser) parseArgs() []*Arg {
 				if p.lex.peek() == '=' {
 					p.lex.consumeToken(tokEqual)
 					defaultValues := []string{}
-					for p.lex.peek() == '[' {
+					if p.lex.peek() == '[' {
 						p.lex.consumeIdent(tokLBracket)
 						for p.lex.peek() != ']' {
 							tex, _ := p.lex.consumeIdentInclString(tokNumber)

--- a/lib/parse.go
+++ b/lib/parse.go
@@ -68,18 +68,18 @@ func (p *Parser) parseArgs() []*Arg {
 				typ, _ := p.lex.consumeIdent()
 				arg.Type = typ.String()
 
-				if p.lex.peek() == '=' {
-					p.lex.consumeToken(tokEqual)
-					tex, _ := p.lex.consumeIdent()
-					te := tex.String()
-					arg.TypeExt = &te
-				}
-
 				if p.lex.peek() == '!' {
 					arg.Null = false
 					p.lex.consumeToken(tokBang)
 				} else {
 					arg.Null = true
+				}
+
+				if p.lex.peek() == '=' {
+					p.lex.consumeToken(tokEqual)
+					tex, _ := p.lex.consumeIdentInclString(tokNumber)
+					te := tex.String()
+					arg.TypeExt = &te
 				}
 			}
 			arg.Directives = p.parseDirectives()

--- a/lib/parse.go
+++ b/lib/parse.go
@@ -64,6 +64,24 @@ func (p *Parser) parseArgs() []*Arg {
 				} else {
 					arg.IsListNull = true
 				}
+
+				if p.lex.peek() == '=' {
+					p.lex.consumeToken(tokEqual)
+					defaultValues := []string{}
+					for p.lex.peek() == '[' {
+						p.lex.consumeIdent(tokLBracket)
+						for p.lex.peek() != ']' {
+							tex, _ := p.lex.consumeIdentInclString(tokNumber)
+							te := tex.String()
+							defaultValues = append(defaultValues, te)
+							if p.lex.peek() == ',' {
+								p.lex.consumeToken(tokComma)
+							}
+						}
+						p.lex.consumeIdent(tokRBracket)
+						arg.DefaultValues = &defaultValues
+					}
+				}
 			} else {
 				typ, _ := p.lex.consumeIdent()
 				arg.Type = typ.String()
@@ -79,7 +97,9 @@ func (p *Parser) parseArgs() []*Arg {
 					p.lex.consumeToken(tokEqual)
 					tex, _ := p.lex.consumeIdentInclString(tokNumber)
 					te := tex.String()
-					arg.DefaultValue = &te
+					defaultValues := []string{}
+					defaultValues = append(defaultValues, te)
+					arg.DefaultValues = &defaultValues
 				}
 			}
 			arg.Directives = p.parseDirectives()

--- a/lib/schema.go
+++ b/lib/schema.go
@@ -330,6 +330,13 @@ func (s *Schema) Parse(p *Parser) {
 					} else {
 						fd.Null = true
 					}
+
+					if p.lex.peek() == '=' {
+						p.lex.consumeToken(tokEqual)
+						tex, _ := p.lex.consumeIdentInclString(tokNumber)
+						te := tex.String()
+						fd.DefaultValue = &te
+					}
 				}
 
 				fd.Directives = p.parseDirectives()

--- a/lib/schema.go
+++ b/lib/schema.go
@@ -319,6 +319,23 @@ func (s *Schema) Parse(p *Parser) {
 					} else {
 						fd.IsListNull = true
 					}
+					if p.lex.peek() == '=' {
+						p.lex.consumeToken(tokEqual)
+						defaultValues := []string{}
+						for p.lex.peek() == '[' {
+							p.lex.consumeIdent(tokLBracket)
+							for p.lex.peek() != ']' {
+								tex, _ := p.lex.consumeIdentInclString(tokNumber)
+								te := tex.String()
+								defaultValues = append(defaultValues, te)
+								if p.lex.peek() == ',' {
+									p.lex.consumeToken(tokComma)
+								}
+							}
+							p.lex.consumeIdent(tokRBracket)
+							fd.DefaultValues = &defaultValues
+						}
+					}
 				} else {
 					fd.IsList = false
 					fd.IsListNull = false
@@ -335,7 +352,9 @@ func (s *Schema) Parse(p *Parser) {
 						p.lex.consumeToken(tokEqual)
 						tex, _ := p.lex.consumeIdentInclString(tokNumber)
 						te := tex.String()
-						fd.DefaultValue = &te
+						defaultValues := []string{}
+						defaultValues = append(defaultValues, te)
+						fd.DefaultValues = &defaultValues
 					}
 				}
 

--- a/lib/schema.go
+++ b/lib/schema.go
@@ -322,7 +322,7 @@ func (s *Schema) Parse(p *Parser) {
 					if p.lex.peek() == '=' {
 						p.lex.consumeToken(tokEqual)
 						defaultValues := []string{}
-						for p.lex.peek() == '[' {
+						if p.lex.peek() == '[' {
 							p.lex.consumeIdent(tokLBracket)
 							for p.lex.peek() != ']' {
 								tex, _ := p.lex.consumeIdentInclString(tokNumber)

--- a/lib/write.go
+++ b/lib/write.go
@@ -281,13 +281,21 @@ func (ms *MergedSchema) stitchArgument(a *Arg, l int, i int) {
 		if !a.IsListNull {
 			ms.buf.WriteString("!")
 		}
+		if a.DefaultValues != nil {
+			ms.buf.WriteString(" = ")
+			ms.buf.WriteString("[")
+			ms.stitchDefaultValues(a.DefaultValues)
+			ms.buf.WriteString("]")
+		}
+		ms.stitchDirectives(a.Directives)
 	} else {
 		ms.buf.WriteString(a.Type)
 		if !a.Null {
 			ms.buf.WriteString("!")
 		}
-		if a.DefaultValue != nil {
-			ms.buf.WriteString(" = " + *a.DefaultValue)
+		if a.DefaultValues != nil {
+			ms.buf.WriteString(" = ")
+			ms.stitchDefaultValues(a.DefaultValues)
 		}
 		ms.stitchDirectives(a.Directives)
 	}
@@ -310,6 +318,17 @@ func (ms *MergedSchema) stitchDirectives(a []*Directive) {
 					ms.stitchDirectiveArgument(b, m, i)
 				}
 				ms.buf.WriteString(")")
+			}
+		}
+	}
+}
+
+func (ms *MergedSchema) stitchDefaultValues(a *[]string) {
+	if l := len(*a); l > 0 {
+		for i, v := range *a {
+			ms.buf.WriteString(v)
+			if i < l-1 {
+				ms.buf.WriteString(", ")
 			}
 		}
 	}

--- a/lib/write.go
+++ b/lib/write.go
@@ -280,12 +280,13 @@ func (ms *MergedSchema) stitchArgument(a *Arg, l int, i int) {
 		}
 	} else {
 		ms.buf.WriteString(a.Type)
-		if a.TypeExt != nil {
-			ms.buf.WriteString(" = " + *a.TypeExt)
-		}
 		if !a.Null {
 			ms.buf.WriteString("!")
 		}
+		if a.TypeExt != nil {
+			ms.buf.WriteString(" = " + *a.TypeExt)
+		}
+		ms.stitchDirectives(a.Directives)
 	}
 
 	if l <= 2 && i != l-1 {

--- a/lib/write.go
+++ b/lib/write.go
@@ -286,8 +286,8 @@ func (ms *MergedSchema) stitchArgument(a *Arg, l int, i int) {
 		if !a.Null {
 			ms.buf.WriteString("!")
 		}
-		if a.TypeExt != nil {
-			ms.buf.WriteString(" = " + *a.TypeExt)
+		if a.DefaultValue != nil {
+			ms.buf.WriteString(" = " + *a.DefaultValue)
 		}
 		ms.stitchDirectives(a.Directives)
 	}

--- a/lib/write.go
+++ b/lib/write.go
@@ -237,10 +237,17 @@ func (ms *MergedSchema) WriteSchema(s *Schema) string {
 			if p.IsList && !p.IsListNull {
 				ms.buf.WriteString("!")
 			}
-			if p.DefaultValue != nil {
-				ms.buf.WriteString(" = " + *p.DefaultValue)
+			if p.DefaultValues != nil {
+				if p.IsList {
+					ms.buf.WriteString(" = ")
+					ms.buf.WriteString("[")
+					ms.stitchDefaultValues(p.DefaultValues)
+					ms.buf.WriteString("]")
+				} else {
+					ms.buf.WriteString(" = ")
+					ms.stitchDefaultValues(p.DefaultValues)
+				}
 			}
-
 			ms.stitchDirectives(p.Directives)
 
 			ms.buf.WriteString("\n")

--- a/lib/write.go
+++ b/lib/write.go
@@ -237,6 +237,9 @@ func (ms *MergedSchema) WriteSchema(s *Schema) string {
 			if p.IsList && !p.IsListNull {
 				ms.buf.WriteString("!")
 			}
+			if p.DefaultValue != nil {
+				ms.buf.WriteString(" = " + *p.DefaultValue)
+			}
 
 			ms.stitchDirectives(p.Directives)
 

--- a/lib/write.go
+++ b/lib/write.go
@@ -1,6 +1,7 @@
 package lib
 
 import (
+	"sort"
 	"strings"
 )
 
@@ -316,6 +317,9 @@ func (ms *MergedSchema) stitchArgument(a *Arg, l int, i int) {
 }
 
 func (ms *MergedSchema) stitchDirectives(a []*Directive) {
+	sort.SliceStable(a, func(i, j int) bool {
+		return a[i].Name > a[j].Name
+	})
 	if l := len(a); l > 0 {
 		for _, a := range a {
 			ms.buf.WriteString(" @" + a.Name)

--- a/test/default_value/generated.graphql
+++ b/test/default_value/generated.graphql
@@ -1,0 +1,12 @@
+type Query {
+    test(X: Int = 20): Int
+    test(X: Int! = 20): Int
+    test1(X: Int = ADMIN): Int
+    test2(X: String! = "user"): Int
+    test3(X: String = "user" @deprecated): Int
+    test3(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
+}
+
+
+
+

--- a/test/default_value/generated.graphql
+++ b/test/default_value/generated.graphql
@@ -10,3 +10,6 @@ type Query {
 
 
 
+input User {
+    name: String! = "woonki"
+}

--- a/test/default_value/generated.graphql
+++ b/test/default_value/generated.graphql
@@ -18,4 +18,5 @@ type Query {
 
 input User {
     name: String! = "woonki"
+    nicknames: [String!]! = ["mununki", "arnold"]
 }

--- a/test/default_value/generated.graphql
+++ b/test/default_value/generated.graphql
@@ -1,10 +1,16 @@
 type Query {
-    test(X: Int = 20): Int
-    test(X: Int! = 20): Int
-    test1(X: Int = ADMIN): Int
-    test2(X: String! = "user"): Int
-    test3(X: String = "user" @deprecated): Int
-    test3(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
+    test1(X: Int = 20): Int
+    test2(X: Int! = 20): Int
+    test3(X: User = ADMIN): Int
+    test4(X: String! = "user"): Int
+    test5(X: String = "user" @deprecated): Int
+    test6(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
+    test7(X: [Int] = [20]): Int
+    test8(X: [Int] = [20, 30]): Int
+    test9(X: [User!]! = [ADMIN]): Int
+    test10(X: [String!] = ["user"]): Int
+    test11(X: [String] = ["user", "user1"] @deprecated): Int
+    test12(X: [String] = ["user", "user1"] @deprecated, Y: String! = "operator" @unique): Int
 }
 
 

--- a/test/default_value/schema/a.graphql
+++ b/test/default_value/schema/a.graphql
@@ -15,4 +15,5 @@ type Query {
 
 input User {
   name: String! = "woonki"
+  nicknames: [String!]! = ["mununki", "arnold"]
 }

--- a/test/default_value/schema/a.graphql
+++ b/test/default_value/schema/a.graphql
@@ -1,0 +1,5 @@
+type Query {
+  test(X: Int = 20): Int
+  test1(X: Int = ADMIN): Int
+  test2(X: String = "user"): Int
+}

--- a/test/default_value/schema/a.graphql
+++ b/test/default_value/schema/a.graphql
@@ -1,10 +1,16 @@
 type Query {
-  test(X: Int = 20): Int
-  test(X: Int! = 20): Int
-  test1(X: Int = ADMIN): Int
-  test2(X: String! = "user"): Int
-  test3(X: String = "user" @deprecated): Int
-  test3(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
+  test1(X: Int = 20): Int
+  test2(X: Int! = 20): Int
+  test3(X: User = ADMIN): Int
+  test4(X: String! = "user"): Int
+  test5(X: String = "user" @deprecated): Int
+  test6(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
+  test7(X: [Int] = [20]): Int
+  test8(X: [Int] = [20, 30]): Int
+  test9(X: [User!]! = [ADMIN]): Int
+  test10(X: [String!] = ["user"]): Int
+  test11(X: [String] = ["user", "user1"] @deprecated): Int
+  test12(X: [String] = ["user", "user1"] @deprecated, Y: String! = "operator" @unique): Int
 }
 
 input User {

--- a/test/default_value/schema/a.graphql
+++ b/test/default_value/schema/a.graphql
@@ -6,3 +6,7 @@ type Query {
   test3(X: String = "user" @deprecated): Int
   test3(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
 }
+
+input User {
+  name: String! = "woonki"
+}

--- a/test/default_value/schema/a.graphql
+++ b/test/default_value/schema/a.graphql
@@ -1,5 +1,8 @@
 type Query {
   test(X: Int = 20): Int
+  test(X: Int! = 20): Int
   test1(X: Int = ADMIN): Int
-  test2(X: String = "user"): Int
+  test2(X: String! = "user"): Int
+  test3(X: String = "user" @deprecated): Int
+  test3(X: String = "user" @deprecated, Y: String! = "operator" @unique): Int
 }

--- a/test/object_extension/generated.graphql
+++ b/test/object_extension/generated.graphql
@@ -1,4 +1,4 @@
-type Person implements Node @talkable @walkable {
+type Person implements Node @walkable @talkable {
     id: ID!
     createTime: Time!
     updateTime: Time!

